### DISCUSSION TITLE
RGB glossary, part I

### DISCUSSION THREAD

**Discussion by dr-orlovsky** [RGB-WG/rgb#52] [General]
## Legal terms

<dl>
<dt>RGB contract</dt><dd>is an agreement between <em>contract parties</em> structured in a form of bearer digital contract. The contract may relate to the both digital of physical assets and events. Contracts which relate to only digital world potentially may be settled in an automatic way without involving any third parties or indermidiaries outside of the <em>contract parties</em>. RGB contracts may provide an ability for the broader public, which is not a part of the <em>contract parties</em> to participate contract operations via mechanism of <em>valencies</em> and <em>state extensions</em>.
<p>A contract has <em>state</em> and business logic, expressed in terms of <em>ownership rights</em> and <em>executive rights</em>. The contract state, rights and conditions of valid <em>operations</em> are defined using <em>RGB schema</em>; only state and operations which allowed by the schema declarations and valdiation scripts are allowed to happen withint the contract scope.</p></dd>

<dt>Contract participant</dt><dd>an actor which participate in <em>contract operations</em>. Contract parties are classified into the following categories:
<ul>
    <li><em>Contract issuer</em>, an actor creating contract <em>genesis</em></li>
    <li><em>Contract party</em> &ndash; all actors which has <em>ownership rights</em> over RGB contract state</li>
    <li>Public &ndash an actor constructing <em>state extensions</em>. Can exist only in contracts providing <em>valencies</em> and <em>state extensions</em>.</li>
</ul></dd>

<dt>Contract issuer</dt><dd>a <em>contract participant</em> has created the <em>contract genesis</em></dd>

<dt>Contract parties</dt><dd>RGB contract always have a well-defined parties of the contract, which hold <em>ownership rights</em> over <em>atoms of state</em>. The state owned by a contract party is called <em>owned state</em>.A contract party is always also a <em>contract participant</em>.
</dd>

<dt>Contract rights</dt><dd>RGB contract parties have a different rights as a part of the contract conditions defined through <em>RGB schema</em>.
The rights under RGB contract can be classified into the following categories:
<ul>
    <li><em>Ownership rights</em></li>
    <li><em>Executive rights</em></li>
    <li><em>Public rights</em></li>
</ul>
</dd>

<dt>Ownership rights</dt><dd>a party having rights over some of the <em>owned contract state</em>. This is exactly the same party which can spend bitcoin UTXO serving as a <em>single-use-seal definition</em> for the state <em>assignment</em>.
<p>It is important to note that ownership rights and ability to spend UTXO participating assignment doesn't necessary implies ability to update the contract state or construct a valid state transition (see <em>executive rights</em>). See <em>rights mismatch</em> section for the details.</p></dd>

<dt>Executive rights</dt><dd>ability to update the contract state in a final form, i.e. to construct a valid <em>state transition</em> satisfying <em>schema</em> validation rules. See also <em>rights mismatch</em> and <em>ownership rights</em> to better understand the difference of executive and ownership rights in the scope of RGB smart contract.</dd>

<dt>Public rights</dt><dd>a right under some schemata to use a contract <em>valency</em> and construct a valid <em>state extensions</em>, which can be later included into the state via reference from a different valid <em>state transition</em>. See <em>state extension</em> for the details.</dd>

<dt>Rights mismatch</dt><dd>a mismatch between <em>ownership rights</em> and <em>executive rights</em>; an inability to provide an update to the state owned by a party. This is caused due to the schema bugs leading to the impossibility to construct a valid <em>state transition</em> with a specific set of ownership rights of the party and/or combination of <em>seal definitions</em> in their <em>assignments</em>.
<p>Rights mismatch may happen in two main cases:</p>
<ul>
    <ol>A user by mistake has assigned different rights to the same UTXO and the schema doesn't provide a way of spending that combination of rights with a valid <em>state transition</em>. This is a reason why properly-designed <em>schema</em> must provide a <em>rights split transition</em> option.</ol>
    <ol>A schema doesn't provide a way of constructing state transition for a defined <em>owned state</em></ol>
</ul>
<p>Contractum compiler notifies about this two bugs during schema compilation, however schemata designed without Contractum may still contain it.</p></dd>
</dl>

## RGB consensus-level terms, part I

<dl>
<dt>RGB consensus</dt><dd>A form of <em>client-side-validation</em> consensus based on bitcoin consensus, UTXO <em>TapRet</em>/<em>OpRet</em> <em>single-use-seals</em>, <em>multi-protocol commitments</em> and <em>consensus-ordered</em> history of <em>contract operations</em> valid under a given <em>RGB schema</em>
<p>RGB consensus has the following layers</p>
<ol>
    <li>Nakamoto consensus (bitcoin proof of work)</li>
    <li><em>LNPBP10 seals</em></li>
    <li><em>Deterministic bitcoin commitments</em> using either <em>TapRet</em> and <em>OpRet</em> procedures</li>
    <li><em>Multi-protocol commitments</em> with LNPBP-4 standard, allowing to pack state transitions under different RGB contracts in an independent way</li>
</ol></dd>

<dt>Data type</dt>
<dd>is a declaration of structured data used in an RGB contract as a part of the <em>state</em>. Uses <em>strict encoding schema</em>, i.e. data types are always composed of strict encoding type primitives and use strict encoding for the binary serialization.
<p>Data types are defined in the same terms as variable types in any structural language (C, Rust, Haskell, etc).</p></dd>

<dt>Contract state</dt>
<dd>the state of an RGB contract; consists of <em>state atoms</em> of different <em>data types</em>. State classified into two categories: <em>global state</em> and <em>owned state</em>.
<p><em>Contract participants</em> usually have only a partial view on the contract state, with some parts of the state not known to them at all (even the fact of its existence), some parts known only in form of zero-knowledge proofs &ndash; and some state known.</p>
<p>The contract state is constructed starting from the contract <em>genesis</em> by consecutively applying <em>operations</em> from the history in their <em>consensus order</em>. Notwithstanding the fact of the partial view of the contract participant (and, as a result, of RGB Node run by that participant) on the contract state, consensus ordering allows the following statements to hold true</p>
<ul>
    <li>Contract state is a part of the RGB consensus</li>
    <li>All participants have a non-contradictory information about the contract state, meaning that if peer `A` has state transitions `a` < `c`, peer `B` has state transitions `a` < `b`, and peer `C` has state transitions `b` < `c`, than their orders are in concord between each other and with a global ordering of `a` < `b` < `c`.</li>
</ul></dd>

<dt>Contract operations</dt>
<dd>An update to the contract state performed according to the rules defined in the contract schema.
<p>Contract operations include:</p>
<ul>
    <li><em>Genesis</em></li>
    <li><em>State transitions</em></li>
    <li><em>State extensions</em></li>
</ul></dd>

<dt>Contract nodes</dt>
<dd>The same as <em>contract operations</em>, but in a serialized on in-memory structured form. Used in the context of RGB core and standard libraries APIs.</dd>

<dt>Consensus order (of contract operations)</dt><dd>Deterministic ordering of all <em>operations</em> under a given contract based on the bitcoin consensus.
<p>All state transitions in an RGB contract (both stash-stored and new one coming with consignment) have a deterministic ordering, which will be the same on the consensus level, meaning that each RGB Node (or other software based on RGB Core-core consensus engine) will come to the same order of state transition for a given contract. If two peers have a different subset of state transitions, the order they get will still be non-contradictory, specifically: if peer `A` has state transitions `a` < `c`, peer `B` has state transitions `a` < `b`, and peer `C` has state transitions `b` < `c`, than their orders are in concord between each other and with a global ordering of `a` < `b` < `c`.</p>
<p>The full definition of the consensus ordering can be found in <a href="https://github.com/RGB-WG/rgb-core/issues/117">RGB core issue #117</a></p></dd>

<dt>RGB schema</dt><dd>The schema defines a rules of how the contract operates. Specifically it declares
<ul>
    <li><em>contract state</em> of the contract</li>
    <li>rules for updating the contract state</li>
</ul>
<p>A schema does that by providing declarations and implementation for the following components:
<ul>
    <li>used data types</li>
    <li>types of <em>global state</em> and <em>owned state</em></li>
    <li>structure of the <em>contract operations</em></li>
    <li>validation logic for the contract state and operations, defined through (1) type system and (2) custom validation scripts, compiled to either <em>EmbeddedVM</em> or <em>AluVM</em> virtual machine instructions</li>
</ul></p>
<p>In the result, a schema defines the following:
<ul>
    <li><em>Global state</em></li>
    <li><em>Ownership rights</em></li>
    <li><em>Executive rights</em></li>
    <li><em>Public rights</em></li>
</ul></p>
<p>One may think of a schema as a "class" definition in the OOP world; in that terms specific RGB contracts are "class instance" created by the schema constructor (see <em>genesis</em>). Such approach allows to separate role of contract developer (in RGB called <em>schema developer</em>) and <em>contract issuer</em>: the first one needs to be an experienced developer, while the second one is not required to know anything about coding or security at all. This also promotes re-use of common codebase by different issuers for the same typical use cases (like fungible assets), reducing the risk of mistakes.</p>
<p>See also <em>RGB consensus</em></p></dd>

<dt>Genesis</dt><dd>a <em>contrat operation</em> initializing the contract state. Must commit to a specific <em>schema</em> under which the contract is created.</dd>

<dt>State transition</dt><dd> is a <em>contract operation</em> updating the <em>contract state</em> which may be performed by a <em>contract party</em> possessing <em>executive rights</em>. A set of state transitions available for the contract is defined in the <em>schema</em>.
<p>A state transition has inputs, referencing the <em>assignments</em> from the previous <em>contract operations</em>, outputs, constructing new <em>assignments</em>, and an optional declaration of <em>valencies</em>.</p></dd>

<dt>State extension</dt><dd>allows to extend (add a new or update existing) contract state. It can can be created by anybody without doing an on-chain commitment (i.e. without closing any <em>single-use-seals</em>). In this it is similar to <em>genesis</em>. However, the state created by the extension is not "final" (compare to an non-mined transaction in the bitcoin mempool) until it gets included into a valid state transition (like transaction gets mined by being included into bitcoin valid block).
<p>Probably the simplest way to understand state extensions is by example. Let's assume we'd like to do an asset which can be issued by anybody in the world through burning equivalent amount of bitcoins. Such operation can't be a state transition, since we do not have a predefined set of single-use-seals to define a "rights of issue" for an open and unknown set of participants. We can't also do "multiple geneses", since each genesis will define a new RGB contract and the assets under different RGB contracts are not fungible. State extensions were created exactly to address this issue.</p></dd>

<dt>Assignment</dt><dd>A one-to-one mapping between <em>seal definition</em> and state data. A state assigned to a single-use-seal is called <em>owned state</em>.</dd>

<dt>Owned state</dt><dd>A form of state s assigned to a <em>seal definition</em> (see also <em>single-use-seals</em>). It is created as a result of <em>contract operations</em> applied in the <em>consensus order</em>, starting from the <em>contract genesis</em>, which defines the initial contract state. The types of the owned state, its validation rules and how contract operations may affect it are defined in the <em>schema</em>.</dd>

<dt>Global state</dt><dd>in many situations a contract needs a state which is "global" for the contract and is not "owned" by anyone (i.e. not assigned to a UTXO <em>single-use-seal</em>, see <em>owned state</em>). For instance, the name of a fungible asset (and ticker) is that sort of information: it does not "belong" to a contract party but instead is a "property" of the contract itself. Another example could be the name of the identity holder which can't be changed and thus can't be assigned to a UTXO single-use-seal.
<p>It is important to note, that while the global state is not "directly owned" by any <em>contract party</em>, it still may be updated through state transitions &ndash; but only if schema enables that explicitly. In such cases the use of the global state (as compared to the owned state) is reasonable only when it may be possible to opt-out from future state updates (i.e. do not define a new single-use-seals) but still needs to keep the state. The asset name is again such an example, since even if the asset can be renamed, future renames after certain renaming may be prohibited (i.e. no new owned state can be created), but the asset name should stay; global allows exactly that.
</p></dd>

<dt>Valency</dt><dd>declaration a possibility of the public participation in the contract logic made in the <em>schema</em> (see <em>public rights</em>). To be executed it requires the schema also to declare a <em>state transition</em>< using the valency./dd>

<dt>Contract interface</dt><dd>a standard for contract API as it can be used by the software, like RGB node, wallets etc. Contract interface provides a way to abstract contract-specific business logic and present pieces of it in a standard form. For instance, it can be used as a way of communicating with RGB Node, like asking it for a semantically-meaningful state and creating operations.
<p>The concept of RGB contract interface is similar to the concept of ERC standards and ABI files in Ethereum world; the most common interfaces are called "RGBxx" and are defined as a separate LNP/BP standards.</p></dd>
<p>List of interfaces standardized by LNP/BP Standards Association:</p>
<ul>
    <li><strong>RGB20</strong>: fungible assets (tokens etc) issued by centralized party</li>
    <li><strong>RGB21</strong>: non-fungible collectibles ("traditional" NFTs). Represent ownership rights for any form of digital or physical property, including art and media files (audio, video, books, texts, images etc).</li>
    <li><strong>RGB22</strong>: digital identity standard (also NFT)</li>
    <li><strong>RGB23</strong>: provable audit logs</li>
    <li><strong>RGB30</strong>: fungible assets with decentralized issuance and on-chain issue proofs (wrapped bitcoins etc)</li>
</ul>

<dt>Rights split transition (RST)</dt><dd>A transition taking any combination of all <em>owned rights</em> defined by a <em>schema</em> and mapping them one-to-one to the new <em>assignments</em>, without changing the contract <em>state</em> or doing any other form of validation. Technically equivalent to the <em>blank transition</em>, but semantically different to it since it is constructed by a wallet in solving <em>right mismatches</em> and not by an RGB Node.</dd>

<dt>Blank transition (BT)</dt><dd>A transition taking any combination of all <em>owned rights</em> defined by a <em>schema</em> and mapping them one-to-one to the new <em>assignments</em>, without changing the contract <em>state</em> or doing any other form of validation. Technically equivalent to the <em>blank transition</em>, but semantically different to it since it is constructed by an RGB Node (and not by a wallet) in moving other contract's state assigned to the same <em>seal</em> that is closed in as a part of a given <em>state transition</em> under some contract.</dd>
</dl>

**Comment by EthanShang8989** [comment]
Could you please add explanations for the following vocabulary: 'stock', 'stash', 'inventory', 'batch', 'opid', 'bundles','bindle','hoard'and 'fascia'?
